overlord/snapstate/snapstate.go

```diff
@@ -1978,10 +1992,94 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, re
 	if len(updateTss.Refresh) > 0 {
 		updateTss.Refresh = finalizeUpdate(st, updateTss.Refresh, len(updates) > 0, updated, userID, flags)
 	}
-
 	return updated, updateTss, nil
 }
 
+// canSplitRefresh returns whether the refresh is a standard refresh of a mix
+// of essential and non-essential snaps on a hybrid system. If the refresh
+// can be split, it also returns the two split update groups.
+func canSplitRefresh(deviceCtx DeviceContext, infos []minimalInstallInfo, flags *Flags) (essential, nonEssential []minimalInstallInfo, split bool) {
+	// TODO: support split refresh for auto-refresh as well
+	if !deviceCtx.IsCoreBoot() || !release.OnClassic || flags.IsAutoRefresh {
+		return nil, nil, false
+	}
+
+	essential, nonEssential = splitEssentialUpdates(infos)
+	if len(essential) == 0 || len(nonEssential) == 0 {
+		return nil, nil, false
+	}
+
+	return essential, nonEssential, true
+}
+
+// splitRefresh creates two separate tasksets for the essential and non-essential
+// snap refresh groups, creating dependencies between specific tasks when
+// appropriate (e.g., snapd is present and should be refreshed first).
+func splitRefresh(essential, nonEssential []minimalInstallInfo, updateFunc func([]minimalInstallInfo) ([]string, *UpdateTaskSets, error)) ([]string, *UpdateTaskSets, error) {
+	// taskset with essential snaps (snapd, kernel and gadget)
+	essentialUpdated, essentialTss, err := updateFunc(essential)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// taskset with non-essential snaps (apps and their bases)
+	nonEssentialUpdated, nonEssentialTss, err := updateFunc(nonEssential)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	allUpdated := append(essentialUpdated, nonEssentialUpdated...)
+
+	// if snapd is in the essential snaps set, the non-essentials must wait for it
```

Comment:
is the waiting on bases taken care now by the prerequisites logic even if the base ends up in the nonEssential but is used by the e.g. the gadget in essentials?

Reply:
My understanding was that in the case of hybrid systems specifically, the base doesn't need to be refreshed before the gadget/kernel because it's not a boot base and it can be refreshed after the reboot

Reply:
it's the the case that a base needs to be refreshed before the snaps using it, that's independent from reboot, but should also be taken care by prereq logic?

Reply:
Is that because in general a snap may have been updated based on a new revision of the base in which case that needs to be refreshed first or is there another reason (asking specifically in the case of a gadget)? 
The prereq logic for kernel and gadgets is a no-op, because the model base would already be installed and if being refreshed it would've been sorted before the gadget. So in that case the commit putting bases in essentialSnaps set needs to be reverted

Reply:
because the base is needed to run hooks, so if we try to run the base and the snap refresh in parallel things might fail, the normal approach is to update bases first (this part doesn't relate to reboots)

Reply:
I've reverted this now plus some simplifications and another test case